### PR TITLE
Switch from `chrono` to `time`

### PR DESCRIPTION
**Pull request by connec**
I'm interested in seeing https://github.com/est31/rcgen/issues/65 resolved, so thought I'd take a crack at replacing `chrono` with `time` here to see if there was appetite to merge such a change.

Overall the change was fairly mechanical, with `time`s API actually being a bit neater in the end, imo. The main limitation seems to be regarding leap seconds. For now I've gone for simply ignoring them. Another option would be to record it in the `UTCTime`/`GeneralizedTime` structs themselves, so that it can be round-tripped in parsing/formatting (but would still not be recorded in the underlying `OffsetDateTime`, exposed through the `datetime()` method).

Since `chrono` types were used in the API, this is a breaking API change. Since a breaking change was already necessary, I also changed the feature to `time`.

I've tried to preserve the existing formatting, however some of it may have drifted to default `rustfmt` style. Let me know if anything needs changed.

### CODE REVIEW

**Comment by connec** [comment]
So it looks like `time ≥ 3` is not compatible with Rust 1.36.0. Furthermore `time` has quite an aggressive [policy](https://github.com/time-rs/time/blob/main/README.md#minimum-rust-version-policy) regarding MSRV:

> The time crate is guaranteed to compile with any release of rustc from the past six months. Optional feature flags that enable interoperability with third-party crates (e.g. rand) follow the policy of that crate if stricter.

Additionally, [one of the change notes](https://github.com/time-rs/time/blob/main/CHANGELOG.md#changed-1) says:

> Per the policy in the README, this may be bumped within the 0.3 series without being a breaking change.

The current MSRV for the latest `time` (`0.3.4`) appears to be 1.51. For the `0.3` line the lowest MSRV is 1.48. It looks like `time ≥ 0.2.23` is also advisory free, and it seems that would let the MSRV drop to 1.32.0.

I don't see any documentation for MSRV for this repo.

The two sensible options seem to be:

- Drop to `time = "0.2.27"` and retain the 1.36.0 MSRV, rather than adopt the comparatively fast-paced MSRV for `time ≥ 3`. This would perhaps be the best option for the stability of this crate, at the expense of potentially missing important updates to `time` (plus, having crate ambiguity w/ use of the latest version).

- Adopt a similar policy as `time` regarding MSRV wrt to optional features, i.e. "Optional feature flags that enable interoperability with third-party crates (e.g. rand) follow the policy of that crate if stricter". This seems like a nice compromise, and is perhaps what people looking to interoperate with `time` would prefer, rather than having to pin their own dependency to `0.2.27`. This may require periodic CI updates to account for the inexorable march of `time`'s MSRV.

**Comment by connec** [comment]
I went ahead and added a commit taking the second approach (caveat the documented MSRV with "Optional feature flags that enable interoperability with third-party crates (e.g. `time`) follow the policy of that crate if stricter"), and updated the CI workflow to not test with features on MSRV.

Let me know if you'd prefer another approach.

File: src/models/time.rs
Comment (est31) [inline src/models/time.rs:192]: Why this change? It makes the code less consistent.

File: src/models/time.rs
Comment (est31) [inline src/models/time.rs:565]: Same here

File: src/models/time.rs
Comment (est31) [inline src/models/time.rs]: This is definitely not great.

**Review by est31 — CHANGES_REQUESTED** [review]
Thanks for filing this PR! Some remarks
* the docs of the `from_datetime` and `from_datetime_opt` and `datetime` functions should be changed to mention `OffsetDateTime` instead of datetime.
* regarding the MSRV, I'm open towards increasing it to as recent as 1.51 to match the time crate, but then ideally yasna.rs itself also uses at least one feature that has been introduced by the respective version. The changes regarding MSRV from this PR are fine, just saying that this can be the subject of future PRs. Increasing the MSRV to older releases than 1.51 is also possible, there just needs to be a feature that is worth the change.
* the biggest issue I have with the PR is the missing handling of leap seconds. They still exist and even single seconds are [treated seriously](https://community.letsencrypt.org/t/2021-06-08-certificate-lifetime-incident-valid-for-an-extra-one-second/153426) in the security domain. I'm not sure I can accept a PR that switches to a library that ignores leap seconds, but happy to hear input on this.

File: src/models/time.rs
Comment (connec) [reply src/models/time.rs:192]: This is because [`time::OffsetDateTime::month`](https://docs.rs/time/0.3.4/time/struct.OffsetDateTime.html#method.month) returns a [`Month`](https://docs.rs/time/0.3.4/time/enum.Month.html) enum variant, so this has to be cast before arithmetic can be performed.

Perhaps, since this one has to change, the rest could be changed to restore consistency.

Comment (est31) [reply src/models/time.rs:192]: Oh then it's okay I guess.

**Comment by connec** [comment]
Thanks for the feedback!

I'll see what can be done regarding leap second handling. I didn't find much when searching the `time` repo for leap seconds but hopefully something will be possible.

Good catch about the docs, I should review them more carefully.

Regarding the MSRV, I think that makes sense. I'm inclined to leave it as-is in this PR, sadly not sure I'm able to go down the route of a more general upgrade.

**Comment by est31** [comment]
Re the leap seconds, I've opened an issue at: https://github.com/time-rs/time/issues/377 . It seems that the time maintainers want proper handling instead of the improper one that the chrono crate has.

File: src/models/time.rs
Comment (est31) [inline src/models/time.rs]: Could you add a TODO here to implement them once time supports them?

**Review by est31 — COMMENTED** [review]
So I've thought about it a little, and it's definitely not great that round trip ability is hurt by this. Should the need arise (users complain or such), we can implement our own datatype and provide converters to the time formst.

I really hope future versions of the time crate support leap seconds and if so, we can enable them. Could you add TODOs to the test as well as to the place where we are ignoring leap seconds?

**Comment by connec** [comment]
I've pushed a couple of commits, one fixing the docs and another with a hack for round-tripping leap seconds. The hack brings back the previous test assertion, and I updated the documentation for `GeneralizedTime::datetime` to indicate that leap seconds are discarded in the returned `OffsetDateTime`.

If you'd prefer I can drop the hack and just document as TODO – I started with this before I noticed your last comment 😄

**Review by est31 — APPROVED** [review]
Thanks!

**Pull request merged**
